Add data size score metric

meljbruno opened the pull request:
Performs statistical analysis on origin connection data sizes and adds a data size score (ds_score) metric to the beacon analysis output.

Zalgo2462 commented on analysis/structure/uconn.go:
```diff
@@ -88,6 +88,9 @@ func getUniqueConnectionsScript(sysCfg *config.SystemConfig) (string, string, []
 						}},
 					}},
 				}},
+				{"orig_bytes", bson.D{
```
This should be removed

meljbruno replied:
Fixed in 0ea6d0536b6373fcb75f0d64cf7cb28ee0a6f2e8.

Zalgo2462 commented on datatypes/structure/structure.go:
```diff
@@ -18,7 +18,7 @@ type (
 		Dst             string        `bson:"dst"`
 		LocalSrc        bool          `bson:"local_src"`
 		LocalDst        bool          `bson:"local_dst"`
-		TotalBytes      int           `bson:"total_bytes"`
```
Undo this change. We need the total bytes associated with unique connections.

meljbruno replied:
Fixed in 0ea6d0536b6373fcb75f0d64cf7cb28ee0a6f2e8.

Zalgo2462 commented on analysis/structure/uconn.go:
```diff
@@ -110,6 +113,7 @@ func getUniqueConnectionsScript(sysCfg *config.SystemConfig) (string, string, []
 				{"local_src", 1},
 				{"local_dst", 1},
 				{"total_bytes", 1},
+				{"orig_bytes", 1},
```
remove this as well

meljbruno replied:
Fixed in 0ea6d0536b6373fcb75f0d64cf7cb28ee0a6f2e8.

Zalgo2462 commented on datatypes/structure/structure.go:
```diff
@@ -18,6 +18,7 @@ type (
 		LocalSrc        bool          `bson:"local_src"`
 		LocalDst        bool          `bson:"local_dst"`
+		TotalBytes      int           `bson:"total_bytes"`
 		OriginBytes     int64         `bson:"origin_bytes"`
```
We'll want to remove this line

Zalgo2462 replied:
Specifically the origin bytes line

meljbruno commented:
Extraneous line in structure.go fixed, score metrics merged, and new data size metrics added in commit 8ee957a9b4368ae355852e9c53097f08dd9d82a5.

Zalgo2462 commented:
Look's good! We should verify that previously highly scored beacons are still highly scored.

Zalgo2462 commented:
Can you please rebase this off of dev before we merge it in? Thanks!

meljbruno commented:
Debasing dev should be all set.

meljbruno commented:
Added the data size metrics to HTML beacon reports, removed "TS" from score variable names and headers in commit 3caa12d1eafae6dd77be9d523d49b970ab85d393.

ethack commented:
@Zalgo2462 Is this good to be merged?

Zalgo2462 commented:
I'm currently getting an error when running html-report.
`html/template:beacon.html: "<" in attribute name: " Mode</th><th>Intvl. Mode Count<"`

Zalgo2462 commented:
`size_mode` and `size_mode_count` appear to be displaying the same information in the html report.

Zalgo2462 commented on analysis/beacon/beacon.go:
```diff
@@ -284,24 +307,40 @@ func (t *Beacon) analyze() {
-		//in order of ascending importance: skew, duration, dispersion
-		output.TS_score = (alpha + beta + gamma) / 3.0
+		//in order of ascending importance: timestamp skew, timestamp duration,
+		//timestamp dispersion, size skew, size duration, size weight
```
s/size duration/size dispersion/

Zalgo2462 commented:
Also, while this code is open for changes, I'd suggest that we remove the TS_duration metric. I've found better results without it. Also, the theory behind TS_duration doesn't make much sense. The metric looks at the observation period (which is usually 1 day), and looks at the time between the first and last connection between the two hosts. It divides this span of time by the observation period. Basically, it tests if a beacon filled the observation period.

meljbruno commented:
Fixed both issues in commit 633215cf5041dbd0a7fd4ba31e352c782dc04d1e. Opening a new pull request for the removal of the TS_duration metric.
